### PR TITLE
Mock SnapshotStore and StoreScanner

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR provides a mocked SnapshotStore and StoreScanner which will produce rows according to given data fixtures. It is the foundation of layered benchmark so that higher layered benchmarks will not involve with RocksDB's performance.

## What are the type of the changes? (mandatory)

- New feature (non-breaking change which adds functionality)

## How has this PR been tested? (mandatory)

New unit tests.

## Benchmark result if necessary (optional)

Currently there is a clone when creating scanner from the mocked store due to unimplemented Rust [GATs](https://github.com/rust-lang/rfcs/blob/master/text/1598-generic_associated_types.md) feature. It is NOT negligible so that when reviewing benchmark results of a higher level, the cost of creating scanner should be taken into consideration. The cost on my machine is shown below:

```
test storage::txn::store::benches::bench_fixture_batch_get    ... bench:      11,825 ns/iter (+/- 3,676)
test storage::txn::store::benches::bench_fixture_get          ... bench:          80 ns/iter (+/- 13)
test storage::txn::store::benches::bench_fixture_scanner      ... bench:     278,860 ns/iter (+/- 114,239)
test storage::txn::store::benches::bench_fixture_scanner_next ... bench:     290,798 ns/iter (+/- 87,633)
test storage::txn::store::benches::bench_fixture_scanner_scan ... bench:     362,404 ns/iter (+/- 118,416)
```

### CODE REVIEW

**Comment by breeswish** [comment]
@solotzg @MyonKeminta PTAL

**Comment by siddontang** [comment]
Very very cool work !!!!!!

File: src/coprocessor/dag/executor/scanner.rs [hunk 1]
```diff
@@ -27,65 +27,59 @@ pub enum ScanOn {
...
-    store.scanner(scan_mode, key_only, lower_bound, upper_bound)
+    store.scanner(asc, key_only, lower_bound, upper_bound)
 }
 
 // `Scanner` is a helper struct to wrap all common scan operations
 // for `TableScanExecutor` and `IndexScanExecutor`
-pub struct Scanner<S: Snapshot> {
-    scan_mode: ScanMode,
+pub struct Scanner<S: Store> {
+    asc: bool,
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/dag/executor/scanner.rs]: good idea

File: src/storage/txn/store.rs [hunk 2]
```diff
@@ -11,13 +11,54 @@
...
 use storage::mvcc::{
     BackwardScanner, BackwardScannerBuilder, ForwardScanner, ForwardScannerBuilder,
 };
 use storage::mvcc::{Error as MvccError, MvccReader};
-use storage::{Key, KvPair, ScanMode, Snapshot, Statistics, Value};
+use storage::{Key, KvPair, Snapshot, Statistics, Value};
+
+use super::{Error, Result};
+
+pub trait Store {
```
Comment (DorianZheng) [inline src/storage/txn/store.rs]: Maybe we can rename it to `StoreReader`? There are too much `Store`.

File: src/storage/txn/store.rs [hunk 3]
```diff
@@ -116,39 +137,40 @@ impl<S: Snapshot> SnapshotStore<S> {
...
+            snapshot,
+            start_ts,
+            isolation_level,
+            fill_cache,
+        }
+    }
+}
+
 pub struct StoreScanner<S: Snapshot> {
     forward_scanner: Option<ForwardScanner<S>>,
```
Comment (DorianZheng) [inline src/storage/txn/store.rs]: Maybe we only need one scanner? They have the same APIs and there behaviors are different and they never exist at the same time

Comment (breeswish) [reply src/storage/txn/store.rs]: Changes are too much. Let's change it in another PR. @MyonKeminta

File: src/storage/txn/store.rs (hunk 2 above)
Comment (breeswish) [reply src/storage/txn/store.rs]: Good idea!

Comment (breeswish) [reply src/storage/txn/store.rs]: Seems that too many places uses `Store`, for example, `SnapshotStore` should be renamed to `SnapshotStoreReader` as well if we would like to change this name. Let's defer it to another PR?

**Pull request merged**